src/lib/integrations/spotify/spotifyAuth.ts

```diff
@@ -0,0 +1,185 @@
+      window.location.href = url
+    } else {
+      await openUrl(url)
+    }
+  }
+
+  static async handleCallback(code: string, state: string | null): Promise<void> {
+    try {
+      
+      // Verify state matches what we stored
+      const storedState = localStorage.getItem(STATE_KEY)
+      
+      if (state === null || state !== storedState) {
+        throw new Error('State mismatch')
+      }
+      
+      // Clear stored state
+      localStorage.removeItem(STATE_KEY)
+
+      const tokens = await this.exchangeCodeForTokens(code)
+      
+      this.setStoredTokens({
+        access_token: tokens.access_token,
+        refresh_token: tokens.refresh_token,
+        expires_at: new Date(Date.now() + tokens.expires_in * 1000).toISOString()
+      })
+
+      // Verify connection by fetching profile
+      const profile = await SpotifyApiService.getUserProfile()
+      if (!profile) {
+        throw new Error('Failed to verify connection')
+      }
+    } catch (error) {
+      console.error('Error in handleCallback:', error)
+      throw error
+    }
+  }
+
+  static async disconnect(): Promise<void> {
+    this.clearStoredTokens()
+  }
+
+  private static async exchangeCodeForTokens(code: string) {
+    const response = await fetch('https://accounts.spotify.com/api/token', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        'Authorization': `Basic ${btoa(`${SPOTIFY_CONFIG.clientId}:${SPOTIFY_CONFIG.clientSecret}`)}`,
+      },
+      body: new URLSearchParams({
+        grant_type: 'authorization_code',
+        code: code,
+        redirect_uri: SPOTIFY_CONFIG.redirectUri,
+      }),
+    })
+
+    if (!response.ok) {
+      const errorText = await response.text()
+      throw new Error(`Failed to exchange code for tokens: ${errorText}`)
+    }
+
+    return response.json()
+  }
+
+  private static async refreshAccessToken(): Promise<SpotifyTokens | undefined> {
+    const tokens = this.getStoredTokens()
+    if (!tokens?.refresh_token) throw new Error('No refresh token available')
+
+    const response = await fetch('https://accounts.spotify.com/api/token', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        Authorization: `Basic ${btoa(`${SPOTIFY_CONFIG.clientId}:${SPOTIFY_CONFIG.clientSecret}`)}`,
+      },
+      body: new URLSearchParams({
+        grant_type: 'refresh_token',
+        refresh_token: tokens.refresh_token,
+      }),
+    })
+
+    if (!response.ok) {
+      console.log('Failed to refresh token', response)
+      throw new Error('Failed to refresh token')
+    }
+
+    const newTokens = await response.json()
+    this.setStoredTokens({
+      access_token: newTokens.access_token,
+      refresh_token: newTokens.refresh_token ?? tokens.refresh_token,
+      expires_at: new Date(Date.now() + newTokens.expires_in * 1000).toISOString()
+    })
+    return newTokens
```

Comment:
This is I think the problem with the refresh logic. It was refreshing the token, but just setting them to local storage wasn't enough. The new tokens need to be returned back